src/rewriter/README.md

```diff
@@ -0,0 +1,231 @@
+a condition that ensure the well-definedness of some expression. The condition
+that the number of repeats in `(repeat n x)` cannot be negative is one such
+example.
+
+It is unnecessary in the example below to enforce `0 <= i <= j < (bvsize x)`.
+The type checker will automatically ensure the expressions entering the
+reconstruction algorithms have sound types.
+
+``` lisp
+(define-rule bv-extract-extract
+  ((x ?BitVec) (i Int) (j Int) (k Int) (l Int))
+  (extract l k (extract j i x))
+  (extract (+ i l) (+ i k) x))
+```
+
+### List Modifier
+
+The `:list` modifier represents an arbitrary number of arguments of the same
+type (gradual types are supported). In the below example, `xs` and `ys` match an
+arbitrary number of boolean arguments.
+
+``` lisp
+(define-rule bool-or-true ((xs Bool :list) (ys Bool :list)) (or xs true ys) true)
+```
+
+### Fixed-Point Rules
+
+The `*` in `define-rule*` indicates that the rule shall be executed by the
+reconstruction algorithm until the expression reaches a fixed point. This is an
+optimisation and useful for writing rules that iterate over the arguments of
+n-ary operators.
+
+Below is an example which uses gradual type and recursively flattens a concat
+using fixed-point rules.
+
+``` lisp
+(define-rule* bv-concat-flatten
+  ((xs ?BitVec :list)
+   (s ?BitVec)
+   (ys ?BitVec :list)
+   (zs ?BitVec :list))
+  (concat xs (concat s ys) zs)
+  (concat xs s ys zs))
+```
+
+The user can optionally supply a rewrite context to support the continuation.
+The context indicates how to use the recursion step, which rewrites *match* to
+*target*, to construct the final result. Below is an example
+
+``` lisp
+(define-rule* str-len-concat-rec ((s1 String) (s2 String) (s3 String :list))
+	(str.len (str.++ s1 s2 s3))
+	(str.len (str.++ s2 s3))
+	(+ (str.len s1) _))
+```
+
+This rule specifies that the rewrite of `(str.len (str.++ s1 s2 ...))` is the
+sum of `(str.len s1)` and the result of applying the rule to the term `(str.len
+(str.++ s2 ...))`
+
+
+## Adding New Operators
+
+This section pertains only to parsing RARE rules in cvc5.
+
+The RARE parser is located at `src/rewriter`. To add a new operator,
+write a new enumeration entry in `src/rewriter/node.py`. The syntax and
+processing are controlled by `rewriter.py` and `parser.py`.
+
+The symbols usable in `(<id> <expr>+)` can be found in `node.py`. The sorts
+available can be found in `mkrewrites.py`.
+
+It is possible in the rewriter to enable parsing of rules of the form
+`define-cond-rule*`. However the condition may interact unexpectedly with the
+fixed-point rule.
+
+```
+<rule> ::= (define-cond-rule* <symbol> (<par>*) [<defs>] <expr> <expr> <expr> [<expr>])
+```
+
+## Changes from Previous Works
+
+In comparison to
+[fmcad22](https://ieeexplore.ieee.org/abstract/document/10026573), we have
+removed the `let` expression.  `let` is replaced by `def` which allow symbols to
+be shared across the condition, match, and target terms.
```

Comment:
```suggestion
Further documentation about the RARE language can be found in [FMCAD 22/Reconstructing Fine-Grained Proofs of Rewrites Using a Domain-Specific Language](https://ieeexplore.ieee.org/document/10026573).  The following changes have been made since the original publication:

`let` expressions were replaced by `def`, which allows symbols to
be shared across the condition, match, and target terms.
```

Reply:
fixed